// smart-fs/lib/logic/smart-write.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const assert = require('assert');

const fs = require('fs');

const path = require('path');

const isEqual = require('lodash.isequal');

const cloneDeep = require('lodash.clonedeep');

const fsExtra = require('fs-extra');

const stringify = require('json-stringify-pretty-compact');

const yaml = require('yaml-boost');

const objectAlign = require('object-align');

const smartRead = require('./smart-read');

const xmlParser = require('../util/xml-parser');

const getExt = require('../util/get-ext');

module.exports = (filepath, content, options = {}) => {
  assert(typeof filepath === 'string');
  assert(content instanceof Object);
  assert(options instanceof Object && !Array.isArray(options));

  const ctx = _objectSpread({
    treatAs: null,
    mergeStrategy: (existing, changeset) => changeset,
    create: true,
    pretty: false,
    keepOrder: true,
    resolve: false
  }, options);

  assert(Object.keys(ctx).length === 6, 'Unexpected Option provided!');
  assert(ctx.treatAs === null || typeof ctx.treatAs === 'string');
  assert(typeof ctx.mergeStrategy === 'function');
  assert(typeof ctx.create === 'boolean');
  assert(typeof ctx.pretty === 'boolean');
  assert(typeof ctx.keepOrder === 'boolean');
  assert(typeof ctx.resolve === 'boolean');
  const targetExists = fs.existsSync(filepath);

  if (ctx.create !== true && !targetExists) {
    return false;
  }

  const ext = getExt(filepath);
  const currentContent = targetExists ? smartRead(filepath, {
    treatAs: ctx.treatAs === null && ext === 'js' ? 'txt' : ctx.treatAs,
    resolve: ctx.resolve
  }) : null;
  const mergedContent = currentContent == null ? content : ctx.mergeStrategy(cloneDeep(currentContent), cloneDeep(content));

  if (!isEqual(currentContent, mergedContent)) {
    fsExtra.ensureDirSync(path.dirname(filepath));

    if (ctx.keepOrder) {
      objectAlign(mergedContent, currentContent);
    }

    let contentString;

    switch (ctx.treatAs || ext) {
      case 'yml':
      case 'yaml':
        contentString = yaml.dump(mergedContent);
        break;

      case 'xml':
        contentString = xmlParser.stringify(mergedContent, options);
        break;

      case 'json':
        contentString = `${ctx.pretty ? stringify(mergedContent) : JSON.stringify(mergedContent, null, 2)}\n`;
        break;

      default:
        assert(Array.isArray(mergedContent));
        contentString = `${mergedContent.join('\n')}\n`;
        break;
    }

    fs.writeFileSync(filepath, contentString);
    return true;
  }

  return false;
};
